aikido_perception/src/AprilTagsModule.cpp

```diff
@@ -0,0 +1,170 @@
+								  std::string referenceFrameId, dart::dynamics::Frame* referenceLink):
+		mNode(node),
+		mMarkerTopic(markerTopic),
+		mConfigData(configData),
+		mResourceRetriever(resourceRetriever),
+		mReferenceFrameId(referenceFrameId),
+		mReferenceLink(referenceLink)
+{
+	mListener = new tf::TransformListener();
+}
+
+//================================================================================================================================
+void AprilTagsModule::detectObjects(std::vector<dart::dynamics::SkeletonPtr>& skeleton_list,double _timeout, ros::Time timestamp)
+{
+	bool transform_lookedup = false;
+	//Looks at all detected tags, looks up config file 
+	//Appends new skeletons to skeleton list
+	do{
+		visualization_msgs::MarkerArrayConstPtr marker_message
+	 			= ros::topic::waitForMessage<visualization_msgs::MarkerArray>(mMarkerTopic,mNode,ros::Duration(_timeout));
+
+		for (const auto& marker_transform : marker_message->markers)
+		{
+			//Get marker, tag ID, and detection transform name 
+			const auto& marker_stamp = marker_transform.header.stamp;
+			const auto& tag_name = marker_transform.ns;
+			const auto& detection_frame = marker_transform.header.frame_id;
+
+			if(marker_stamp < timestamp){
+				dtwarn << "Marker stamp " << marker_stamp<< " for " << tag_name 
+				<< "is behind timestamp parameter: " <<timestamp <<"\n";
+
+				continue;
+			}
+
+			std::string skel_name;
+			Eigen::Isometry3d skel_offset;
+			dart::common::Uri skel_resource;
+
+			//check if tag name is in database
+			if (mConfigData->getTagNameOffset(tag_name,skel_name,skel_resource,skel_offset))
+			{
+
+				//Get orientation of marker
+				Eigen::Isometry3d marker_pose = aikido::perception::convertROSPoseToEigen(marker_transform.pose);
+
+
+				//For the frame-frame transform
+				tf::StampedTransform transform;
+
+				try{
+					mListener->waitForTransform(mReferenceFrameId,detection_frame,
+						marker_stamp,ros::Duration(_timeout));
+
+					mListener->lookupTransform(mReferenceFrameId,detection_frame,
+						marker_stamp, transform);
+				}
+				catch(const tf::ExtrapolationException& ex){
+					ROS_ERROR("%s",ex.what());
+					continue;
+				}
+				catch(const tf::TransformException& ex){
+					ROS_ERROR("%s",ex.what());
+					throw std::runtime_error("TF Error which cannot be recovered");
+				}
+
+				transform_lookedup = true;
+
+				//Get the transform as a pose matrix
+				Eigen::Isometry3d frame_pose = aikido::perception::convertStampedTransformToEigen(transform);
+
+				//Compose to get actual skeleton pose
+				Eigen::Isometry3d temp_pose = frame_pose * marker_pose;
+				Eigen::Isometry3d skel_pose = temp_pose * skel_offset;
+
+				Eigen::Isometry3d link_offset = mReferenceLink->getWorldTransform();
+				skel_pose = link_offset * skel_pose;
+
+				//Check if skel in skel_list
+				//If there then update pose
+				//If not then append skeleton
+				skel_name.append(std::to_string(marker_transform.id));
```

Comment:
What is the logic behind appending the `id` here? Could `getTagNameOffset` do this internally?